TextToSpeech/impl/TTSURLConstructer.cpp

```diff
@@ -23,106 +23,171 @@ TTSURLConstructer::~TTSURLConstructer()
+
+    // Sanitize String
+    std::string sanitizedString;
+    if(isfallback)
+    {
+        sanitizeString(config.getFallbackValue(), sanitizedString);
+    }
+    else
+    {
+        sanitizeString(text, sanitizedString);
+    }
+    tts_request.append("&text=");
+    tts_request.append(sanitizedString);
     return tts_request;
 }
 
+std::string  TTSURLConstructer::httppostURL(TTSConfiguration &config,std::string text,bool isfallback)
+{  
+   std::string tts_request;
+   CURL *curl;
+   CURLcode res;
+   struct curl_slist *list = NULL;
+   std::string readBuffer;
+   curl = curl_easy_init();
+   if(curl)
+   {
+       JsonObject jsonConfig;
+       JsonObject parameters;
+       std::string post_data;
+       if(isfallback)
+       {
+          jsonConfig["input"] = config.getFallbackValue();
+       }
+       else
+       {
+          jsonConfig["input"] = text;
+       }
+       jsonConfig["language"] = config.language();
+       jsonConfig["voice"] = config.voice();
+       jsonConfig["encoding"] = "mp3";
+       jsonConfig.ToString(post_data);
+       TTSLOG_INFO("gcd postdata :%s\n",post_data.c_str());
+
+       curl_easy_setopt(curl, CURLOPT_URL,config.secureEndPoint().c_str());
+       curl_easy_setopt(curl, CURLOPT_POST, 1L);
+       curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 2L);
+       curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_data.c_str());
+       list = curl_slist_append(list, "content-type: application/json");
+       list = curl_slist_append(list, (std::string("x-api-key: ") +
+                                            config.apiKey()).c_str() );
+       curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list);
+       curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+       curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
+
+       res = curl_easy_perform(curl);
+	   if ( res != CURLE_OK )
+       {
+          TTSLOG_ERROR("TTS: Error in interacting with endpoint. CURL error is:  %s\n", curl_easy_strerror(res));
+          if( config.isFallbackEnabled() && isfallback == false)
+          {
+             TTSLOG_INFO("RDK TTS: Device is not connected with Internet, hence speaking fallback text in place of actual text: %s",text.c_str());
+             tts_request.assign(config.getFallbackPath());
+          }
+       }
+       else
+       {
+          TTSLOG_INFO("gcd curl response :%s\n",readBuffer.c_str());
+          parameters.FromString(readBuffer);
+          tts_request.assign(parameters["url"].String());
+       }
+       curl_slist_free_all(list);
+       curl_easy_cleanup(curl);
+	}  
+        return tts_request; 
+
+}
+
+std::string TTSURLConstructer::constructURL(TTSConfiguration &config,std::string text,bool isfallback)
+{
+     // EndPoint URL
+    //std::string tts_request;
+    if(!(config.apiKey().empty()))
+    {
+       if(config.isSwitchRequired())
+       {
+          if(WPEFramework::Plugin::TTS::NetworkStatusObserver::getNetworkStatusObserver()->isConnected())
```

Comment:
Detecting network state & determining whether the local endpoint should be used, shouldn't be here. It should be with TTSSpeaker. This way, this TTSURLConstructor's scope will just be constructing the URL based on configuration passed. We can determine which URL should be used from TTSSpeaker & pass it to constructor (this would also help to consider the m_remoteError case)

bool TTSSpeaker::shouldUseLocalEndpoint() {
    if(hasValidLocalEndpoint())
        return !NetworkObserver::getInstance()->isConnected() || m_remoteError;
    return false;
}

// From wherever we need the URL constructed
urlConstructor.constructURL(config, text, isFallback, shouldUseLocalEndpoint());